packages/yarn-plugin-allow-scripts/test/index.spec.mts

```diff
@@ -0,0 +1,238 @@
+import test from 'ava'
+import { existsSync, rmSync } from 'node:fs'
+import { dirname, join, normalize } from 'node:path'
+import { SpawnSyncOptions } from 'node:child_process'
+import { spawnSync } from 'node:child_process'
+import { npmRunPath } from 'npm-run-path';
+
+const isWindows = process.platform === 'win32';
+const portablePath = isWindows
+  ? (p) => p.replace(/^[\\]/, '')
+  : (p) => p
+
+/**
+ * For fat fingers
+ */
+const PACKAGE_JSON = 'package.json'
+
+const cleanup = (projectName: string) => {
+  const projectRoot = join(dirname(import.meta.url.replace(/^file:/, '')), 'projects', projectName)
+  // post-test cleanup of messy `yarn init`
+  spawnSync(
+    'corepack',
+    [
+      'npm@9',
+      'exec',
+      'rimraf',
+      '.git',
+      '.yarn',
+    ],
+    {
+      cwd: projectRoot,
+      encoding: 'utf-8',
+      shell: isWindows,
+    },
+  );
+  spawnSync(
+    'git',
+    [
+      'clean',
+      '-fdx',
+      normalize(`test/projects/${projectName}`),
+    ],
+    {
+      encoding: 'utf-8',
+      shell: isWindows,
+    },
+  );
+  spawnSync(
+    'git',
+    [
+      'restore',
+      normalize(`test/projects/${projectName}/`),
+    ],
+    {
+      encoding: 'utf-8',
+      shell: isWindows,
+    },
+  );
+};
+
+/**
+ * Execute allow-scripts binaries with the given arguments
+ *
+ * @param t Ava test object
+ * @param args Arguments to pass
+ * @param cwd Working directory to execute command from
+ * @returns Process result
+ */
+const run = (
+  t: any,
+  args: string[],
+  cwd: string,
+  env: {} = {},
+): {exitCode: number|null, out: string, err: string, output: (string|Buffer|null)[], signal: string|null, error?: Error|undefined} => {
+  try {
+    const defaultOpts = realisticEnvOptions(cwd);
+    const result = spawnSync(
+      isWindows
+        ? process.env.COMSPEC || 'cmd.exe'
+        : process.env.SHELL || '/bin/sh',
+      [
+        ...(isWindows ? ['/d', '/s', '/c'] : ['-c']),
+        [
+          isWindows ? 'corepack.cmd' : 'corepack',
+          'yarn@4',
+          ...args
+        ].join(' '),
+      ],
+      {
+        ...defaultOpts,
+        env: {
+          ...process.env,
+          ...defaultOpts.env,
+          ...env,
+        }
+      },
+    )
+    return {
+      exitCode: result.status,
+      out: result?.stdout?.toString()?.trim(),
+      err: result?.stderr?.toString()?.trim(),
+      output: result?.output,
+      signal: result.signal,
+      error: result.error,
```

Comment:
kudos for not forgetting the `error` field

You could rethrow it if it's truthy to fall through to error handling below.